cmd/keymasterd/adminHandlers.go

```diff
@@ -164,54 +168,80 @@ func (state *RuntimeState) deleteUserHandler(
 	case "text/html":
 		http.Redirect(w, r, usersPath, http.StatusFound)
 	default:
-		w.WriteHeader(200)
+		w.WriteHeader(http.StatusOK)
 		fmt.Fprintf(w, "OK\n")
 	}
 }
 
-func (state *RuntimeState) generateBootstrapOTP(
-	w http.ResponseWriter, r *http.Request) {
+func (state *RuntimeState) generateBootstrapOTP(w http.ResponseWriter,
+	r *http.Request) {
 	failure, authUser := state.sendFailureToClientIfNonAdmin(w, r)
 	if failure {
 		return
 	}
-	//TODO: check method, must be POST
 	username := state.ensurePostAndGetUsername(w, r)
 	if username == "" {
 		return
 	}
-
 	profile, existing, fromCache, err := state.LoadUserProfile(username)
 	if err != nil {
-		logger.Printf("error parsing err=%s", err)
-		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		state.logger.Printf("error parsing err=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError,
+			err.Error())
 		return
 	}
 	if !existing {
-		state.writeFailureResponse(w, r, http.StatusBadRequest, "User does not exist in DB")
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"User does not exist in DB")
 		return
 	}
 	if fromCache {
-		state.writeFailureResponse(w, r, http.StatusServiceUnavailable, "Working in db disconnected mode, try again later")
+		state.writeFailureResponse(w, r, http.StatusServiceUnavailable,
+			"Working in db disconnected mode, try again later")
 		return
 	}
-	logger.Printf("profile=%v", profile)
-	bootstrapOTP := bootstrapOTPData{
-		ExpiresAt: time.Now().Add(defaultBootstrapOTPDuration),
+	state.logger.Debugf(1, "profile=%v", profile)
+	duration := defaultBootstrapOTPDuration
+	formDuration, ok := r.Form["duration"]
+	if ok {
+		if len(formDuration) != 1 {
+			state.writeFailureResponse(w, r, http.StatusBadRequest,
+				"Single value duration required")
+			return
+		}
+		duration, err = time.ParseDuration(formDuration[0])
+		if err != nil {
+			state.writeFailureResponse(w, r, http.StatusBadRequest, err.Error())
+			return
+		}
+	}
+	if duration < time.Minute {
+		duration = time.Minute
 	}
+	if duration > time.Hour*24 {
+		state.writeFailureResponse(w, r, http.StatusBadRequest,
+			"Duration over 1 day not allowed")
+		return
+	}
+	bootstrapOTP := bootstrapOTPData{ExpiresAt: time.Now().Add(duration)}
 	bootstrapOTP.Value, err = genRandomString()
 	if err != nil {
-		logger.Printf("error generating randr=%s", err)
-		state.writeFailureResponse(w, r, http.StatusInternalServerError, "")
+		state.logger.Printf("error generating randr=%s", err)
+		state.writeFailureResponse(w, r, http.StatusInternalServerError,
```

Comment:
We never send the actual error on internal errors

Reply:
Fixed.